contracts/truefi/TrueRatingAgencyV2.sol

```diff
@@ -0,0 +1,485 @@
+        if (loanStatus == LoanStatus.Pending) {
+            loans[id].prediction[choice] = loans[id].prediction[choice].sub(stake);
+        }
+
+        // if loan status passed pending state claim TRU reward
+        if (loanStatus >= LoanStatus.Running) {
+            claim(id, msg.sender);
+        }
+
+        // update account votes
+        loans[id].votes[msg.sender][choice] = loans[id].votes[msg.sender][choice].sub(stake);
+
+        // transfer tokens to sender and emit event
+        require(stakedTrustToken.transfer(msg.sender, amountToTransfer));
+        emit Withdrawn(id, msg.sender, stake, amountToTransfer, burned);
+    }
+
+    /**
+     * @dev Internal view to convert values to 8 decimals precision
+     * @param input Value to convert to TRU precision
+     * @return output TRU amount
+     */
+    function toTrustToken(uint256 input) internal pure returns (uint256 output) {
+        output = input.div(TOKEN_PRECISION_DIFFERENCE);
+    }
+
+    /**
+     * @dev Update total TRU reward for a Loan
+     * Reward is divided proportionally based on # TRU staked
+     * chi = (TRU remaining in distributor) / (Total TRU allocated for distribution)
+     * interest = (loan APY * term * principal)
+     * R = Total Reward = (interest * chi * rewardFactor)
+     * @param id Loan ID
+     */
+    modifier calculateTotalReward(address id) {
+        if (loans[id].reward == 0) {
+            uint256 interest = ILoanToken(id).profit();
+
+            // calculate reward
+            // prettier-ignore
+            uint256 totalReward = toTrustToken(
+                interest
+                    .mul(distributor.remaining())
+                    .mul(rewardMultiplier)
+                    .div(distributor.amount())
+            );
+
+            uint256 ratersReward = totalReward.mul(ratersRewardFactor).div(10000);
+            loans[id].reward = ratersReward;
+            if (loans[id].reward > 0) {
+                distributor.distribute(totalReward);
+                trustToken.transfer(address(stakedTrustToken), totalReward.sub(ratersReward));
+            }
+        }
+        _;
+    }
+
+    /**
+     * @dev Claim TRU rewards for voters
+     * - Only can claim TRU rewards for funded loans
+     * - Voters can claim a portion of their total rewards over time
+     * - Claimed automatically when a user withdraws stake
+     *
+     * chi = (TRU remaining in distributor) / (Total TRU allocated for distribution)
+     * interest = (loan APY * term * principal)
+     * R = Total Reward = (interest * chi)
+     * R is distributed to voters based on their proportion of votes/total_votes
+     *
+     * Claimable reward = R x (current time / total time)
+     *      * (account TRU staked / total TRU staked) - (amount claimed)
+     *
+     * @param id Loan ID
+     * @param voter Voter account
+     */
+    function claim(address id, address voter) public override onlyFundedLoans(id) calculateTotalReward(id) {
```

Comment:
I think we can either remove the claim function, or call it from the withdraw function. Does it make sense to have this?

Reply:
We call it from withdraw function. I kept it public, so it's technically possible to claim without withdrawing, but we con't need to implement any UI endpoint for that. All rewards will be automatically claimed with the withdrawal anyway.